statistics/semantic/src/main/java/com/spotify/heroic/statistics/semantic/SemanticMetricBackendReporter.java

```diff
@@ -118,19 +131,34 @@ public MetricBackend decorate(final MetricBackend backend) {
     @Override
     public DataInMemoryReporter newDataInMemoryReporter() {
         return new DataInMemoryReporter() {
-            private final AtomicLong dataRead = new AtomicLong(0);
-            private final AtomicLong dataInMemory = new AtomicLong(0);
-            private final AtomicLong rowsAccessed = new AtomicLong(0);
+            private final LongAccumulator dataRead = new LongAccumulator((x, y) -> x + y, 0L);
+            private final AtomicLong dataInMemory = new AtomicLong(0L);
+            private final LongAccumulator rowsAccessed = new LongAccumulator((x, y) -> x + y, 0L);
+            private final LongAccumulator maxDataInMemory = new LongAccumulator(Long::max, 0L);
+            private final Stopwatch queryWatch = Stopwatch.createStarted();
+            private final Map<Integer, ReadRangeStatistics> seriesDensity =
+                new ConcurrentHashMap<>();
 
             @Override
             public void reportRowsAccessed(final long n) {
-                rowsAccessed.addAndGet(n);
+                rowsAccessed.accumulate(n);
+            }
+
+            @Override
+            public void reportSliceRead(
+                final Series series, final DateRange range, final long n
+            ) {
+                int key = series.hashCode() + range.hashCode();
```

Comment:
Hm, not sure if collisions are acceptable (10k-20k gives 1-5% for me). They would skew up the measurements. If you're concerned about key size, you can use the [`Series#getHashCode()`](https://github.com/spotify/heroic/blob/master/heroic-component/src/main/java/com/spotify/heroic/common/Series.java#L82) which is a 128-bit hash and has a low probability of collisions.